src/compiler/checker.ts

```diff
@@ -3140,7 +3198,32 @@ namespace ts {
                         }
                         // falls through
                     case SyntaxKind.ModuleDeclaration:
-                        if (result = callback(getSymbolOfNode(location as ModuleDeclaration).exports!)) {
+                        const sym = getSymbolOfNode(location as ModuleDeclaration);
+                        // `sym` may not have exports if this module declaration is backed by the symbol for a `const` that's being rewritten
+                        // into a namespace - in such cases, it's best to just let the namespace appear empty (the const members couldn't have referred
+                        // to one another anyway)
+                        if (result = callback(sym.exports || emptySymbols)) {
+                            return result;
+                        }
+                        break;
+                    case SyntaxKind.ClassDeclaration:
+                    case SyntaxKind.ClassExpression:
+                    case SyntaxKind.InterfaceDeclaration:
+                        // Type parameters are bound into `members` lists so they can merge across declarations
+                        // This is troublesome, since in all other respects, they behave like locals :cries:
+                        // TODO: the below is shared with similar code in `resolveName` - in fact, rephrasing all this symbol

```

Comment:
still not a fan of TODOs in source, at least phrased as such

Reply:
Good, then maybe one day one of us will feel like actually spending the time and effort *do*ing it, in order to remove said TODO.